### PR TITLE
Fixing potential null dereferences

### PR DESCRIPTION
**Pull request by tpltnt**
Hi there,

I fixed a few potential NULL pointer dereferences in the way pointed out in #1800.
Cheers,
tpltnt

### CODE REVIEW

**Comment by rocallahan** [comment]
This looks good, except that instead of checking that one call to `malloc`, I'd rather introduce a malloc wrapper into `util.h`/`util.cc`, say `xmalloc`, that aborts on OOM, and replace `malloc` calls with calls to that wrapper. Do you want to do that, or shall I?

**Comment by tpltnt** [comment]
@rocallahan done that. I implemented ```xmalloc``` and replaced all ```malloc``` calls.

File: src/GdbConnection.cc [hunk 1]
```diff
@@ -537,6 +537,8 @@ static string read_target_desc(const char* file_name) {
   string path = exe_directory() + "../share/rr/" + string(file_name);
   stringstream ss;
   FILE* f = fopen(path.c_str(), "r");
+  assert(NULL != f);
```
Comment (rocallahan) [inline src/GdbConnection.cc]: Just `assert(f)`

File: src/log.cc [hunk 2]
```diff
@@ -86,6 +86,8 @@ static void init_log_globals() {
   char* env = getenv(log_env);
   if (env) {
     env = strdup(env);
+    assert(NULL != env);
```
Comment (rocallahan) [inline src/log.cc]: Just `assert(env)`

File: src/preload/preload.c [hunk 3]
```diff
@@ -736,6 +736,8 @@ static void* thread_trampoline(void* arg) {
 int pthread_create(pthread_t* thread, const pthread_attr_t* attr,
                    void* (*start_routine)(void*), void* arg) {
   struct thread_func_data* data = malloc(sizeof(*data));
+  assert(NULL != data);
```
Comment (rocallahan) [inline src/preload/preload.c]: Don't change this here, just make this `xmalloc` later

File: src/util.cc [hunk 4]
```diff
@@ -821,4 +821,12 @@ void copy_file(Task* t, int dest_fd, int src_fd) {
   }
 }
 
+void *xmalloc(size_t size) {
+  void *mem_ptr = malloc(size);
+  if (NULL == mem_ptr) {
```
Comment (rocallahan) [inline src/util.cc]: `if (!mem_ptr)`

File: src/test/rrutil.h [hunk 5]
```diff
@@ -106,6 +106,18 @@ typedef unsigned char uint8_t;
...
+ * Allocate new memory of |size| in bytes. The pointer returned is never NULL.
+ * This calls aborts the program if the host runs out of memory.
+ */
+void *xmalloc(size_t size) {
+  void *mem_ptr = malloc(size);
+  if (NULL == mem_ptr) {
+    abort();
+  }
+  return mem_ptr;
+}
```
Comment (rocallahan) [inline src/test/rrutil.h]: Squash this into the previous patch.  Every prefix of the list of commits should build and run.

File: src/test/rrutil.h [hunk 6]
```diff
@@ -176,7 +176,7 @@ static uint64_t GUARD_VALUE = 0xdeadbeeff00dbaad;
  */
 inline static void* allocate_guard(size_t size, char value) {
   char* cp =
-      (char*)malloc(size + 2 * sizeof(GUARD_VALUE)) + sizeof(GUARD_VALUE);
+      (char*)xmalloc(size + 2 * sizeof(GUARD_VALUE)) + sizeof(GUARD_VALUE);
```
Comment (rocallahan) [inline src/test/rrutil.h]: How do these test changes build, given that tests don't link to `util.cc`? I think you need to declare `xmalloc` in rrutil.h too.

Comment (tpltnt) [inline src/test/rrutil.h]: I copied the xmalloc() implementation to rrutil.h, pretty much at the beginning of the file.

File: src/test/rrutil.h (hunk 5 above)
Comment (tpltnt) [inline src/test/rrutil.h]: I am doing this right now.

File: src/GdbConnection.cc (hunk 1 above)
Comment (tpltnt) [inline src/GdbConnection.cc]: done

File: src/log.cc (hunk 2 above)
Comment (tpltnt) [inline src/log.cc]: done

File: src/preload/preload.c (hunk 3 above)
Comment (tpltnt) [inline src/preload/preload.c]: I already changed this on my local branch.

File: src/util.cc (hunk 4 above)
Comment (tpltnt) [inline src/util.cc]: done

**Comment by tpltnt** [comment]
@rocallahan I fixed all issues you pointed out.

**Comment by rocallahan** [comment]
This looks good overall, but it's still the case that some patches don't build without later patches. Please squash or reorder patches so that each patch (plus its predecessors) builds. If you prefer, just squash them all together.

**Comment by tpltnt** [comment]
I messed up syncing with current master and squashing the commits. New attempt with #1837.

**Pull request closed without merging**